### ISSUE TITLE
src/CppUTest/CMakeLists.txt is not a complete CMake list file

### ISSUE THREAD

**Issue by smfelsher** [cpputest/cpputest#1187]
File `src/CppUTest/CMakeLists.txt` requires the following lines added to it in order to make it a complete CMakeLists file:

```
cmake_minimum_required(VERSION 3.1 FATAL_ERROR)

project(CppUTest LANGUAGES CXX)
```

**Comment by KevinWMatthews** [comment]
These are defined at lines 1 and 8 of the root `CMakeLists.txt`:
https://github.com/cpputest/cpputest/blob/8f9a0130e41cb3d5100e5f05fab7a3d35465c2f9/CMakeLists.txt#L1
https://github.com/cpputest/cpputest/blob/8f9a0130e41cb3d5100e5f05fab7a3d35465c2f9/CMakeLists.txt#L8

What is the benefit of redefining them in a subdirectory?

**Comment by smfelsher** [comment]
According to the instructions in [src/CppUTest/CMakeIntegration-README.md](https://github.com/cpputest/cpputest/blob/master/src/CppUTest/CMakeIntegration-README.md), I can `add_subdirectory` the `src/CppUTest` directory. If I follow the instructions provided, I do not have a complete CMakeLists file for the `src/CppUTest` directory. In addition to the two (2) lines I indicate above, I also had to add the compiler definitions and disable the `install` commands (I didn't take the time to figure out the correct way since I had no desire to install the libraries on my system in this particular instance).

Bottom line, the CMakeLists files for the entire project are not implemented according to CMake best practices: https://www.slideshare.net/DanielPfeifer1/effective-cmake

I do appreciate the fact that this is an open source project and that I can make my proposed changes and submit a pull-request. I may do that in the near future when I have some time to devote to it. For now, at least this open ticket captures the issue.

**Comment by basvodde** [comment]
Thanks @smfelsher . Looking forward to your contribution.

**Comment by KevinWMatthews** [comment]
Hi @smfelsher ,

I see; thanks for the explanation.

Does your setup allow you to call `add_subdirectory()` on CppUTest's root directory:
```cmake
add_subdirectory(${CppUTestRootDirectory} CppUTest)
```
instead of the `src/CppUTest` subdirectory?
```cmake
add_subdirectory(${CppUTestRootDirectory}/src/CppUTest CppUTest)
```

That seems to clear up the issue for me, but I'm only one data point. I'm guessing that the call to `project()` sets `CMAKE_INSTALL_LIBDIR` internally.

It seems like calling [project()](https://cmake.org/cmake/help/v3.1/command/project.html) in a subdirectory would change the CMake variables:

  * `PROJECT_SOURCE_DIR`
  * `PROJECT_BINARY_DIR`

wouldn't it? If so, I wonder if that would mess up standard builds (an untested theory).

Not to discourage a contribution, of course!

**Comment by KevinWMatthews** [comment]
Speaking of CMake being out of date, I wonder if there are alternative approaches to including CppUTest in a project. I haven't tried them yet, but the book *Modern CMake* on GitLab lists two approaches:

  * [git submodules](https://cliutils.gitlab.io/modern-cmake/chapters/projects/submodule.html)
  * [FetchContent](https://cliutils.gitlab.io/modern-cmake/chapters/projects/fetch.html) - CMake 3.11+

Both ideas are intriguing but they may not apply to your specific situation.

Cheers!

**Comment by smfelsher** [comment]
@KevinWMatthews, yes, adding the root directory as a subdirectory in my top-level project does seem to work. So, is there any need for [src/CppUTest/CMakeIntegration-README.md](https://github.com/cpputest/cpputest/blob/master/src/CppUTest/CMakeIntegration-README.md)? It seems to me that this just added to the confusion.

**Comment by KevinWMatthews** [comment]
Good question. There is some reasoning behind why the README was originally written that way - #1102 . Perhaps we can modify the file to meet both sets of needs? I would defer to @basvodde  and @rgrover (see also #1150 and #1186 ) before making a premature fix.

**Comment by basvodde** [comment]
So far, I've left the maintenance of the CMake build to other people as I usually use the autotools build myself. So whatever pull request seems reasonable, I'll accept. Perhaps wait a couple days for the other CMake people to respond.

**Comment by bneumann** [comment]
Hi, I got a similar problem with using cpputest as a subproject. I formerly tried googletest but using gmock in a C only project is quite cumbersome. Anyways the usual approach of building cmake under MinGW (clone and build) works pretty good. But I want to hook it into my CMake project. 

First I added a CMakeCppUTest.cmake.in file which only downloads the repo:

```cmake
cmake_minimum_required(VERSION 2.8.2)

project(cpputest-download NONE)

include(ExternalProject)
ExternalProject_Add(CppUTest
    GIT_REPOSITORY https://github.com/cpputest/cpputest.git
    GIT_TAG master
    SOURCE_DIR "${CMAKE_BINARY_DIR}/cpputest-src"
    BINARY_DIR "${CMAKE_BINARY_DIR}/cpputest-build"
    CONFIGURE_COMMAND ""
    BUILD_COMMAND ""
)
```

Then I put in these lines in my main CMake project:

```cmake
# Download and unpack CppUTest at configure time
configure_file(cmake/CMakeCppUTest.cmake.in cpputest-download/CMakeLists.txt)
execute_process(COMMAND "${CMAKE_COMMAND}" -G "${CMAKE_GENERATOR}" .
    WORKING_DIRECTORY "${CMAKE_BINARY_DIR}/cpputest-download" )
execute_process(COMMAND "${CMAKE_COMMAND}" --build .
    WORKING_DIRECTORY "${CMAKE_BINARY_DIR}/cpputest-download" )
    
set(CppUTestRootDirectory "${CMAKE_BINARY_DIR}/cpputest-build")
# Add CppUTest directly to our build. 
add_subdirectory("${CMAKE_BINARY_DIR}/cpputest-src"
                 "${CMAKE_BINARY_DIR}/cpputest-build")
``` 

And If I try to compile it now I get an error because cmake somehow trys to find the generates folder inside the cppunit-src folder.

```cmd
PS D:\Firmware\D12_Dev\FW\build> make CppUTest
[  0%] Building CXX object cpputest-build/src/CppUTest/CMakeFiles/CppUTest.dir/CommandLineArguments.cpp.obj
In file included from <built-in>:1:
In file included from D:/Firmware/D12_Dev/FW/build/cpputest-src/include/CppUTest/MemoryLeakDetectorNewMacros.h:22:
In file included from D:/Firmware/D12_Dev/FW/build/cpputest-src/include/CppUTest/CppUTestConfig.h:32:
D:/Firmware/D12_Dev/FW/build/cpputest-src/include/CppUTest/CppUTestGeneratedConfig.h:45:10: fatal error: 'generated/CppUTestGeneratedConfig.h' file not found
#include "generated/CppUTestGeneratedConfig.h"
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
make[3]: *** [cpputest-build/src/CppUTest/CMakeFiles/CppUTest.dir/build.make:64: cpputest-build/src/CppUTest/CMakeFiles/CppUTest.dir/CommandLineArguments.cpp.obj] Error 1
make[2]: *** [CMakeFiles/Makefile2:923: cpputest-build/src/CppUTest/CMakeFiles/CppUTest.dir/all] Error 2
make[1]: *** [CMakeFiles/Makefile2:930: cpputest-build/src/CppUTest/CMakeFiles/CppUTest.dir/rule] Error 2
make: *** [Makefile:529: CppUTest] Error 2
```

Anyone got a tip for me what might be the problem?

**Comment by bneumann** [comment]
I think I fixed my problem above. Don't know if that is a valid/lasting solution. I added ```target_include_directories(CppUTest PRIVATE ${CppUTestRootDirectory})``` to my main CMakeLists file. It looks like so now:

```cmake
# Download and unpack CppUTest at configure time
configure_file(cmake/CMakeCppUTest.cmake.in cpputest-download/CMakeLists.txt)
execute_process(COMMAND "${CMAKE_COMMAND}" -G "${CMAKE_GENERATOR}" .
    WORKING_DIRECTORY "${CMAKE_BINARY_DIR}/cpputest-download" )
execute_process(COMMAND "${CMAKE_COMMAND}" --build .
    WORKING_DIRECTORY "${CMAKE_BINARY_DIR}/cpputest-download" )
    
# Do not compile CppUTests own tests    
set(TESTS OFF CACHE BOOL "" FORCE)
# Suppress CMAKE legacy warning
set(CMAKE_LEGACY_CYGWIN_WIN32 1) 
set(CppUTestRootDirectory "${CMAKE_BINARY_DIR}/cpputest-build" CACHE BOOL "" FORCE)
# Add CppUTest directly to our build. 
# Prevent install:
add_subdirectory("${CMAKE_BINARY_DIR}/cpputest-src"
                 "${CMAKE_BINARY_DIR}/cpputest-build" EXCLUDE_FROM_ALL)

target_include_directories(CppUTest PRIVATE ${CppUTestRootDirectory})
```
I also prevent install of the CppUTest stuff, suppress the legacy warning and do not include the internal tests.

**Comment by basvodde** [comment]
Seems solved. Closing.
